### ISSUE TITLE
macOS client doesn't work through SOCKS5 Proxy

### ISSUE THREAD

**Issue by Char1es4k** [samschott/maestral#453]
**Describe the bug**

I'm using dropbox in China, where the service is blocked by national firewall. Dropbox client can work with self created SOCKS5 proxy on rented server. However Maestral client can't make connection the same way, even if I enable SOCKS5-Proxy in WLAN settings or route the client through Proxifier. 
![CleanShot 2021-09-10 at 19 17 39](https://user-images.githubusercontent.com/58471920/132845641-5bc03321-d5a4-4cc2-9048-59216a3fbfbb.png)
With VPN enabled
![CleanShot 2021-09-10 at 19 18 42](https://user-images.githubusercontent.com/58471920/132845709-c9878948-43d8-42b5-9a07-e2da856620ed.png)
VPN disabled, only SOCKS5 proxy.
With conventional VPN Maestral can work fine though, which is a bit strange. 

**To Reproduce**

**Expected behaviour**

Is there any workaround? I tried to convert socks5 to https and route Maestral through https proxy in Proxifier. Sadly it still won't work.
**System:**

 - Maestral version: 1.4.8
 - Python version: 3.9.6
 - OS: macOS 11.5.2
 - Desktop environment: 
 - PyQt version (for Linux GUI):

**Additional context**

**Comment by charles0316** [comment]
I found that I forgot to mosaic email address. And it seems impossible to delete the issue. So I guess I'll leave it open then.

**Comment by SamSchott** [comment]
You can edit your post to delete the screenshot!

**Comment by SamSchott** [comment]
Alternatively, I can delete the issue completely, if you prefer.

**Comment by charles0316** [comment]
Thanks for the advice. Screenshots have been replaced.

**Comment by SamSchott** [comment]
Regarding the actual issue, I'm not sure how SOCKS5 proxies work exactly. Since #434, proxy settings are retrieved from environment variables and handled by the `requests` module. From the `requests` documentation [here](https://docs.python-requests.org/en/master/user/advanced/?highlight=http_proxy#proxies):

> Requests relies on the proxy configuration defined by standard environment variables `http_proxy`, `https_proxy`, `no_proxy` and `curl_ca_bundle`. Uppercase variants of these variables are also supported.

Could you check if the environment variable `https_proxy` is set in your case? I am not sure if macOS will set it for you when you configure a proxy in System Preferences. If it does not, I'll need to look into alternative methods of proxy discovery.

**Comment by charles0316** [comment]
It sounds kinda dumb but I really don't know how to check it(noob in coding).  But I do find out that me using privoxy to forward HTTP to SOCKS5 proxy is only useful for other devices that connect to the same network as my Mac, but not for Mac apps in this every same device. Tried localhost:8118(8118 is default port selection in privoxy), 0.0.0.0:8118, 127.0.0.1:8118, and local IP of my Mac:8118. None of them worked.

I guess I'll have to dig deeper into Privoxy manuals, or resort to other apps for HTTP proxy in this local machine. Maestral sync  should work afterwards.

**Comment by SamSchott** [comment]
> It sounds kinda dumb but I really don't know how to check it(noob in coding). 

Not dumb at all! On macOS, you can check this by opening the Terminal app and typing the command (and hitting enter):
```
echo $https_proxy
```
I suspect that this will just return an empty line, which means that the proxy configuration is not stored in the environment variable but somewhere else.

If it's not too much trouble, you can also try the following: Again in the Terminal app, type `python3` to start Python. Then run the following two commands:
```
import urllib.request
print(urllib.request.getproxies())
```
This time, I hope that the correct proxy configuration should be shown, namely:
```
{'https': '127.0.0.1:8118'}
```
If yes, I'll know how to fix it...

**Comment by charles0316** [comment]
Thanks for detailed instructions. I do get an empty line after the first command. And in Python3, what I have after those 2 commands is 127.0.0.1:1083, which is the original SOCKS5 proxy that I deploy. Unfortunately I didn't take screenshot, cuz I thought what you meant by "correct proxy configuration"was to type echo $https_proxy again, so I just close the terminal and open a new one. It should be: {'socks5': '127.0.0.1:1083'} if I remember it correctly.

**Comment by SamSchott** [comment]
Ok, thanks for the detailed infos.

It appears that the situation is more complicated than I thought. Python's request library will retrieve the proxy settings from the `all_proxy=socks5://127.0.0.1:1083` environment variable which you have already defined for homebrew. If and only if it cannot find any proxy settings in environment variables, it will query proxy settings directly from System Preferences. I suspect that this will be the case for Maestral, since it is a standalone program which is not started from the Terminal. I'm therefore a bit confused why it doesn't just work out of the box.

How and where did you configure the SOCKS5 proxy? Could you also check the return value from the following Python command:
```Python
import urllib.request
print(urllib.request.getproxies_macosx_sysconf())
```
This will bypass all environment settings and directly get the proxy configuration from System Preferences. What does this return in your case?

**Comment by charles0316** [comment]
![CleanShot 2021-09-15 at 12 30 04@2x](https://user-images.githubusercontent.com/58471920/133371308-40b8c174-07fb-4585-b2a8-504b12c05458.png)
It seems that proxy configuration is blank from system preferences.

**Comment by hronro** [comment]
The reliability of `urllib.request.getproxies()` in retrieving proxy information from the System Configuration for macOS and Windows Systems Registry for Windows seems questionable. Below are my settings as displayed in the macOS System Configuration panel:

<img width="827" alt="CleanShot 2023-05-04 at 12 34 11@2x" src="https://user-images.githubusercontent.com/15059605/236113408-6ffec5ec-72f5-465c-9ca1-c2f11837187f.png">

This is what I did using `urllib.request.getproxies()`:

<img width="682" alt="CleanShot 2023-05-04 at 12 34 17@2x" src="https://user-images.githubusercontent.com/15059605/236113440-a543647a-2266-499e-9f3a-cd70c57646b9.png">

I also found some similar issues on Windows:

https://stackoverflow.com/questions/65931275/python-requests-module-does-it-use-system-level-on-windows-proxy-settings
https://github.com/python/cpython/issues/86793

A workaround on macOS is to use `scutil --proxy` to get the proxy information:

<img width="682" alt="CleanShot 2023-05-04 at 12 34 25@2x" src="https://user-images.githubusercontent.com/15059605/236114438-4b7c1b82-880d-4990-95b0-1992972e452e.png">

**Comment by samschott** [comment]
Indeed, it looks like `urllib.request.getproxies()` ignores SOCKS proxies on macOS. See here for you proxies are read from the system configuration:

https://github.com/python/cpython/blob/f5c38382f9c40f0017cef086896a8160e313ac9e/Modules/_scproxy.c#L234

In particular, kSCPropNetProxiesSOCKSEnable and kSCPropNetProxiesSOCKSProxy are never checked. That being said, requests also requires PySocks to be installed. This is not currently bundled in the Maestral macOS app either (will be easy to include though).

I'm a bit reluctant to introduce custom support for SOCKS proxies in Maestral, IMO this should be handled by using an external library, or ideally by allowing Python's urllib to detect them.

**Comment by samschott** [comment]
I've filed https://github.com/python/cpython/issues/104180 upstream and will create PR to fix this in urllib.

**Comment by samschott** [comment]
The fix from https://github.com/python/cpython/pull/104181 included in Python 3.12 and the latest macOS app bundle v1.9.1 is based on Python 3.12.
